### PR TITLE
bpo-34013: Move the Python 2 hints from the exception constructor to the parser

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue34013

### CODE REVIEW

**Comment by pablogsal** [comment]
CC: @ammaraskar this was the PR I had prepared. If you want, we could explore incorporate your suggestions here

**Comment by ammaraskar** [comment]
Nice: image

----

My thinking for suggesting `Did you mean print(...)` (not sure if the one with `end=' '` is worth the added rule) is that this is the type of mistake which beginners are likely to make while blindly following an outdated tutorial or copying a snippet. For those cases, it's almost better to be verbose and show close to what change they should make rather than just say "missing parentheses" which might be too jargony.

**Comment by pablogsal** [comment]
> My thinking for suggesting `Did you mean print(...)` (not sure if the one with `end=' '` is worth the added rule) is that this is the type of mistake which beginners are likely to make while blindly following an outdated tutorial or copying a snippet. For those cases, it's almost better to be verbose and show close to what change they should make rather than just say "missing parentheses" which might be too jargony.

I am a bit skeptical but I trust your juzguement. I have updated the PR with the changes you propose :)

File: Grammar/python.gram [hunk 1]
```diff
@@ -848,9 +848,10 @@ expression_without_invalid[expr_ty]:
     | disjunction
     | lambdef
 invalid_legacy_expression:
-    | a=NAME b=expression_without_invalid {
-        _PyPegen_check_legacy_stmt(p, a) ? RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "Missing parentheses in call to '%U'.", a->v.Name.id) : NULL}
- 
+    | a=NAME b=star_expressions {
+        _PyPegen_check_legacy_stmt(p, a) ? RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b,
```
Comment (ammaraskar) [inline Grammar/python.gram:852]: Where do the carets for this end up? If they're at the end of the expression, LGTM.

**Review by ammaraskar — APPROVED** [review]

File: Grammar/python.gram (hunk 1 above)
Comment (pablogsal) [reply Grammar/python.gram:852]:
```
>>> print a+b, c+d, f()
  File "<stdin>", line 1
    print a+b, c+d, f()
    ^^^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?

>>> 3,4 , [print 3+4, 6+7, 8], 34
  File "<stdin>", line 1
    3,4 , [print 3+4, 6+7, 8], 34
           ^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?

>>> print {3+4, 6+7, 8}, 45, 6
  File "<stdin>", line 1
    print {3+4, 6+7, 8}, 45, 6
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?

>>> (print {3+4, 6+7, 8}, 45, 6), 5
  File "<stdin>", line 1
    (print {3+4, 6+7, 8}, 45, 6), 5
     ^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
```

Comment (ammaraskar) [reply Grammar/python.gram:852]: perfect 👌

**Pull request merged**